notes/inheritance-1.md

```diff
@@ -0,0 +1,329 @@
+```C++
+int main(){
+    Animal a;
+    Bun b;
+    KingBun kb;
+    printRank(a); // prints 1
+    printRank(b); // prints 5
+    printRank(kb); // prints 10
+}
+
+```
+- Yay, code reuse! This is a lot cleaner and satisfies our needs
+
+---
+## Derived-to-base conversions
+
+Before we move on, let's first address some important intuition behind inheritance:
+
+> We can use an object of a derived type as if it were an object of its base type
+- Also called derived-to-base conversion
+
+<br>
+
+Let me give an intuitive explanation for why we can perform derived-to-base conversion.
+
+Think of a derived object as containing multiple parts
+1. A subobject contaiing data members inherited from its base class
+2. A subobject containing the members defined in the derived class
+
+Example:
+
+![](../images/inheritance-subobject.PNG)
+
+Since a derived object contains a subpart corresponding to its base class, we can safely use an object of a derived type as if it were an object of its base type
+- Conceptual example: if you needed a Car to drive somewhere, a Tesla would do just fine. Sure, it contains extra functionality specific to a Tesla, but it also has all the properties of a driveable car
+
+---
+## Implications of derived-to-base conversions
+
+Derived-to-base conversion gives us 2 important facts:
+
+> First, we can bind a pointer/reference (to a base-class type) to an object of a type derived from that base class
+
+This is why, in the previous chapter, we could do something like this:
+
+```C++
+void printRank(Animal &ani){
+    std::cout << ani.Rank() << std::endl;
+}
+
+int main(){
+    Animal a;
+    Bun b;
+    KingBun kb;
+    printRank(a); // prints 1
+    printRank(b); // prints 5
+    printRank(kb); // prints 10
+}
+```
+- Since `Bun` and `KingBun` are all derived classes of `Animal`, we can bind a reference (to `Animal`) to an object of type `Bun` or `KingBun`
+
+<br>
+
+Secondly, since we can do this type of binding, when we use a reference or pointer to a base-class type, we don't know the actual type of the object to which the pointer/reference is bound!
+- The object can be an object of the base class or one of its derived classes
+- In the example above, is `ani` bound to an `Animal`, `Bun`, or `KingBun` object? We don't know until runtime!
+
+---
+## Static type vs. Dynamic type
+
+The **static type** of an expression is always known at compile time
+- It is the type with which a variable is declared or that an expression yields
+
+For example, in our `printRank` function, parameter `ani` has static type `Animal&`
+
+<br>
+
+The **dynamic type** is the type of the object in memory that the variable or expression actually represents
+
+The dynamic type of an variable that isn't a reference or pointer is just the same as the variable's static type
+
+Example:
+```C++
+Bun b;
+```
+- Variable `b`'s static type is `Bun`, its dynamic type is also `Bun`
+- Another way to put it: a variable of type `Bun` is always a `Bun` object
+
```

Comment:
Maybe also list another example, like how on line 7 from 2 code segments before, `ani` is statically an `Animal`, but it is dynamically a `KingBun`.